core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,532 @@
+	*num_rxpkts = j;
+}
+
+static void pl022_rx16(struct spi_chip *chip, uint16_t *rdat,
+	uint32_t *num_rxpkts)
+{
+	uint32_t j = 0;
+	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE)
+			rdat[j++] = read16(pd->base + SSPDR);
+
+	} while (read32(pd->base + SSPSR) & SSPSR_BSY);
+
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+	*num_rxpkts = j;
+}
+
+static void pl022_print_peri_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 22 10 #4 0");
+	DMSG("Read: 0x %x %x %x %x\n",
+		read32(pd->base + SSPPeriphID0),
+		read32(pd->base + SSPPeriphID1),
+		read32(pd->base + SSPPeriphID2),
+		read32(pd->base + SSPPeriphID3));
+}
+
+static void pl022_print_cell_id(struct pl022_data *pd __maybe_unused)
+{
+	DMSG("Expected: 0x 0D F0 05 B1");
+	DMSG("Read: 0x %x %x %x %x\n",
+		read32(pd->base + SSPPCellID0),
+		read32(pd->base + SSPPCellID1),
+		read32(pd->base + SSPPCellID2),
+		read32(pd->base + SSPPCellID3));
+}
+
+static void pl022_sanity_check(struct pl022_data *pd)
+{
+	DMSG("SSPB2BTRANS: Expected: 0x2. Read: 0x%x\n",
+		read32(pd->base + SSPB2BTRANS));
+	pl022_print_peri_id(pd);
+	pl022_print_cell_id(pd);
+}
+
+static inline uint32_t pl022_calc_freq(struct pl022_data *pd,
+	uint8_t cpsdvr, uint8_t scr)
+{
+	return pd->clk_hz / (cpsdvr * (1 + scr));
+}
+
+static void pl022_calc_clk_divisors(struct pl022_data *pd,
+	uint8_t *cpsdvr, uint8_t *scr)
+{
+	uint32_t freq1 = 0, freq2 = 0;
+	uint8_t tmp_cpsdvr1, tmp_scr1, tmp_cpsdvr2 = 0, tmp_scr2 = 0;
+
+	for (tmp_scr1 = SSP_SCR_MIN; tmp_scr1 < SSP_SCR_MAX; tmp_scr1++) {
+		for (tmp_cpsdvr1 = SSP_CPSDVR_MIN; tmp_cpsdvr1 < SSP_CPSDVR_MAX;
+			tmp_cpsdvr1++) {
+			freq1 = pl022_calc_freq(pd, tmp_cpsdvr1, tmp_scr1);
+			if (freq1 == pd->speed_hz)
+				goto done;
+			else if (freq1 < pd->speed_hz)
+				goto stage2;
+		}
+	}
+
+stage2:
+	for (tmp_cpsdvr2 = SSP_CPSDVR_MIN; tmp_cpsdvr2 < SSP_CPSDVR_MAX;
+		tmp_cpsdvr2++) {
+		for (tmp_scr2 = SSP_SCR_MIN; tmp_scr2 < SSP_SCR_MAX;
+			tmp_scr2++) {
+			freq2 = pl022_calc_freq(pd, tmp_cpsdvr2, tmp_scr2);
+			if (freq2 <= pd->speed_hz)
+				goto done;
+		}
+	}
+
+done:
+	if (freq1 >= freq2) {
+		*cpsdvr = tmp_cpsdvr1;
+		*scr = tmp_scr1;
+		DMSG("speed: requested: %u, closest1: %u\n",
+			pd->speed_hz, freq1);
+	} else {
+		*cpsdvr = tmp_cpsdvr2;
+		*scr = tmp_scr2;
+		DMSG("speed: requested: %u, closest2: %u\n",
+			pd->speed_hz, freq2);
+	}
+	DMSG("cpsdvr: %u (0x%x), scr: %u (0x%x)\n",
+		*cpsdvr, *cpsdvr, *scr, *scr);
+}
+
+static void pl022_flush_fifo(struct pl022_data *pd)
+{
+	uint32_t rdat;
+
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE) {
+			rdat = read32(pd->base + SSPDR);
+			DMSG("rdat: 0x%x\n", rdat);
+		}
+	} while (read32(pd->base + SSPSR) & SSPSR_BSY);
+}
+
+void pl022_configure(struct pl022_data *pd)
+{
+	uint16_t mode, data_size;
+	uint8_t cpsdvr, scr, lbm;
+
+	pl022_sanity_check(pd);
```

Comment:
ok to assert all inputs.
But at least, config error cases for `mode` and `data_size_bits` switch/case blocks should call `panic()` instead of `EMSG()`.